### PR TITLE
Change the command

### PR DESCRIPTION
**Pull request by devillove084**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@
     // box_pointers,
     // elided_lifetimes_in_paths, // allow anonymous lifetime
     // missing_copy_implementations, // Copy may cause unnecessary memory copy
-    missing_debug_implementations,
+    // missing_debug_implementations,
```
Comment (pzheng1025) [inline src/main.rs:11]: Don't skip

File: src/main.rs [hunk 2]
```diff
@@ -103,205 +101,230 @@ pub enum RunAsRole {
...
+    Local,
+}
+
+/// Async fuse args type
+#[derive(Debug)]
+pub struct AsyncFuseArgs {
+    /// Node id
     pub node_id: String,
-    /// Node IP
+    /// IP address
```
Comment (pzheng1025) [inline src/main.rs:120]: Keep "Node IP"

File: src/main.rs [hunk 3]
```diff
@@ -402,117 +475,184 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+        let driver_name = get_driver_name(matches);
+        let controller_server = csi::build_grpc_controller_server(
+            &end_point,
+            &driver_name,
+            Arc::<MetaData>::clone(&md),
+        )?;
+        csi::run_grpc_servers(&mut [controller_server], async_server);
+    }
+
+    if let Some(matches) = matches.subcommand_matches("start_node") {
```
Comment (pzheng1025) [inline src/main.rs:578]: use if .. else if ...

File: src/main.rs [hunk 4]
```diff
@@ -103,205 +101,230 @@ pub enum RunAsRole {
...
-        .get_matches();
-    get_args(&matches)
+impl ArgParam {
+    /// Get the arg name for index
+    pub fn get_name(&self) -> &'static str {
+        return self.name;
+    }
+
+    /// Generate the arg with short&long or not
+    pub fn new_arg<'a>(&self) -> Arg<'a, 'a> {
```
Comment (pzheng1025) [inline src/main.rs:160]: Please rewrite this function to reduce duplicate code.

File: src/main.rs [hunk 5]
```diff
@@ -402,117 +475,184 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
-                scheduler_extender.start();
-            });
-            scheduler_extender_thread
-                .join()
-                .unwrap_or_else(|e| panic!("scheduler extender error: {:?}", e));
-        }
-        RunAsRole::AsyncFuse => {
-            if let Err(e) = async_fuse::start_async_fuse(etcd_delegate_clone, &args_clone) {
+    let matches = parse_args();
+    if let Some(matches) = matches.subcommand_matches("start_csi_controller") {
```
Comment (pzheng1025) [inline src/main.rs:563]: Define all subcommands as const string

Comment (devillove084) [reply src/main.rs]: Due to clap, match subcommand must use original str.

File: src/main.rs (hunk 4 above)
Comment (devillove084) [reply src/main.rs:162]: Done

File: src/main.rs (hunk 3 above)
Comment (devillove084) [reply src/main.rs]: Done

File: src/main.rs (hunk 2 above)
Comment (devillove084) [reply src/main.rs]: Done

File: src/main.rs (hunk 1 above)
Comment (devillove084) [reply src/main.rs:11]: Cause some errors.

File: src/main.rs [hunk 6]
```diff
@@ -103,205 +101,214 @@ pub enum RunAsRole {
...
+    /// Get the arg name for index
+    #[must_use]
+    pub const fn get_name(&self) -> &'static str {
+        self.name
+    }
+
+    /// Generate the arg with short&long or not
+    #[must_use]
+    pub fn new_arg<'a>(&self) -> Arg<'a, 'a> {
+        let mut pre = Arg::with_name(self.name)
```
Comment (pzheng1025) [inline src/main.rs:163]: what is the meaning of pre? using `arg` is better?

File: src/main.rs [hunk 7]
```diff
@@ -103,205 +101,214 @@ pub enum RunAsRole {
...
+
+        pre
+    }
 }
 
-/// Get arguments value
-#[allow(clippy::too_many_lines)] //allow for this function as there is no other logic in this function
-fn get_args(matches: &ArgMatches) -> CliArgs {
+/// Generate the default arg
+#[allow(clippy::too_many_lines)]
```
Comment (pzheng1025) [inline src/main.rs:182]: add comment for this allowance. 
For example 
#[allow(clippy::too_many_lines)] //allow for this function as there is no other logic in this function

File: src/main.rs [hunk 8]
```diff
@@ -402,89 +457,140 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
-    )?;
-    let md = Arc::new(meta_data);
-    let async_server = true;
 
-    match args.run_as {
-        RunAsRole::Controller => {
+    let matches = parse_args();
+
+    match matches.subcommand() {
+        ("start_csi_controller", Some(matches)) => {
```
Comment (pzheng1025) [inline src/main.rs:550]: Define const str for different subcommands.

File: src/main.rs [hunk 9]
```diff
@@ -402,89 +457,140 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+        etcd_delegate,
+    )
+}
+
+/// Get command from default map
+fn get_map<'a>(map: &HashMap<&str, Arg<'a, 'a>>, name: &str) -> Arg<'a, 'a> {
+    if let Some(res) = map.get(name) {
+        res.clone()
+    } else {
+        panic!("Get default command fail");
```
Comment (pzheng1025) [inline src/main.rs:501]: provide more information like panic!("Failed to get default argument {}", name);

File: src/main.rs (hunk 8 above)
Comment (devillove084) [reply src/main.rs:550]: The 'subcommand' do not allow define const str replace the first param.

File: src/main.rs (hunk 9 above)
Comment (devillove084) [reply src/main.rs:501]: Done

File: src/main.rs (hunk 7 above)
Comment (devillove084) [reply src/main.rs:182]: Done

File: src/main.rs (hunk 6 above)
Comment (devillove084) [reply src/main.rs:163]: Done

File: src/main.rs [hunk 10]
```diff
@@ -402,89 +458,140 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+    SubCommand::with_name(name)
+        .arg(get_map(&arg_map, MOUNT_POINT_ARG_NAME))
+        .arg(get_map(&arg_map, WORKER_PORT_ARG_NAME))
+        .arg(get_map(&arg_map, NODE_ID_ARG_NAME))
+        .arg(get_map(&arg_map, NODE_IP_ARG_NAME))
+        .arg(get_map(&arg_map, ETCD_ADDRESS_ARG_NAME))
+}
+
+/// Parse the metadata
+pub fn parse_metadata(
```
Comment (pzheng1025) [inline src/main.rs:477]: remove pub

File: src/main.rs [hunk 11]
```diff
@@ -402,89 +458,140 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
         None => VolumeType::Local,
     };
-    CliArgs {
-        end_point,
+    volume_type
+}
+
+/// Generate the metadata
+#[must_use]
+pub fn gen_metadata<'a>(name: &str) -> App<'a, 'a> {
```
Comment (pzheng1025) [inline src/main.rs:466]: remove pub

File: src/main.rs [hunk 12]
```diff
@@ -366,30 +395,57 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
             .collect(),
         None => Vec::new(),
     };
+    etcd_address_vec
+}
+
+/// Get scheduler port
+#[must_use]
+pub fn get_scheduler_port(matches: &ArgMatches) -> u16 {
```
Comment (pzheng1025) [inline src/main.rs:403]: remove pub, also for other local functions

Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 11 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 10 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 1 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs [hunk 13]
```diff
@@ -103,205 +109,215 @@ pub enum RunAsRole {
...
+        if let Some(l) = self.long {
+            arg = arg.long(l);
+        }
+
+        arg
+    }
+}
+
+/// Generate the default arg
+//allow for this function as there is no other logic in this function
```
Comment (pzheng1025) [inline src/main.rs:190]: put this comment after #[allow(clippy::too_many_lines)]

File: src/main.rs [hunk 14]
```diff
@@ -402,89 +466,142 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+        .subcommand(
+            gen_metadata("start_csi_controller")
+                .arg(get_map(&arg_map, DRIVER_NAME_ARG_NAME))
+                .arg(get_map(&arg_map, END_POINT_ARG_NAME)),
+        )
+        .subcommand(
+            gen_metadata("start_node")
+                .arg(get_map(&arg_map, DRIVER_NAME_ARG_NAME))
+                .arg(get_map(&arg_map, END_POINT_ARG_NAME))
+                // Now gen async fuse data other args
```
Comment (pzheng1025) [inline src/main.rs:527]: This comment doesn't make sense

File: src/main.rs [hunk 15]
```diff
@@ -402,89 +466,142 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+                .arg(get_map(&arg_map, MOUNT_POINT_ARG_NAME))
+                .arg(get_map(&arg_map, ETCD_ADDRESS_ARG_NAME)),
+        )
+        .get_matches();
+    matches
+}
+
+// TODO: The pattern_type_mismatch is false positive
+// issue link: `https://github.com/rust-lang/rust-clippy/issues/7946`
+#[allow(clippy::too_many_lines, clippy::pattern_type_mismatch)]
```
Comment (pzheng1025) [inline src/main.rs:554]: Move `#[allow(clippy::pattern_type_mismatch)]` right above `match matches.subcommand() {`

File: src/main.rs [hunk 16]
```diff
@@ -495,10 +612,16 @@ fn main() -> anyhow::Result<()> {
                 .join()
                 .unwrap_or_else(|e| panic!("csi thread error: {:?}", e));
         }
-        RunAsRole::SchedulerExtender => {
+        (START_SCHEDULER_EXTENDER, Some(matches)) => {
+            let metadata = parse_metadata(matches, RunAsRole::SchedulerExtender)?;
+            let md = Arc::new(metadata);
+            let sep = get_scheduler_port(matches);
```
Comment (pzheng1025) [inline src/main.rs:618]: change `sep` to `port` or `scheduler_port`. sep is not a good name.

File: src/main.rs (hunk 13 above)
Comment (devillove084) [reply src/main.rs:190]: done

File: src/main.rs (hunk 15 above)
Comment (devillove084) [reply src/main.rs:554]: done

File: src/main.rs (hunk 14 above)
Comment (devillove084) [reply src/main.rs:527]: done

File: src/main.rs (hunk 16 above)
Comment (devillove084) [reply src/main.rs:618]: done

File: src/main.rs [hunk 17]
```diff
@@ -402,89 +466,143 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+        panic!("Failed to get default argument {}", name);
     }
 }
 
+/// Parse the args in subcommand
+fn parse_args() -> ArgMatches<'static> {
+    let arg_map = get_default_arg_map();
+    let matches = App::new("datenlord")
+        .subcommand(
+            gen_metadata("start_csi_controller")
```
Comment (pzheng1025) [inline src/main.rs:519]: change to `START_CONTROLLER`

File: src/main.rs [hunk 18]
```diff
@@ -402,89 +466,143 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+                .arg(get_map(&arg_map, DRIVER_NAME_ARG_NAME))
+                .arg(get_map(&arg_map, END_POINT_ARG_NAME)),
+        )
+        .subcommand(
+            gen_metadata("start_node")
```
Comment (pzheng1025) [inline src/main.rs:524]: change this too.

File: src/main.rs [hunk 19]
```diff
@@ -402,89 +466,143 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+                .arg(get_map(&arg_map, DRIVER_NAME_ARG_NAME))
+                .arg(get_map(&arg_map, END_POINT_ARG_NAME))
+                .arg(get_map(&arg_map, SERVER_PORT_NUM_ARG_NAME))
+                .arg(get_map(&arg_map, VOLUME_TYPE_ARG_NAME))
+                .arg(get_map(&arg_map, CACHE_CAPACITY_ARG_NAME))
+                .arg(get_map(&arg_map, VOLUME_INFO_ARG_NAME)),
+        )
+        .subcommand(
+            gen_metadata("start_scheduler_extender")
```
Comment (pzheng1025) [inline src/main.rs:533]: change this too.

File: src/main.rs [hunk 20]
```diff
@@ -402,89 +466,143 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+                .arg(get_map(&arg_map, SCHEDULER_EXTENDER_PORT_ARG_NAME)),
+        )
+        .subcommand(
+            SubCommand::with_name("async_fuse")
```
Comment (pzheng1025) [inline src/main.rs:537]: change this too.

Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 19 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 18 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs (hunk 17 above)
Comment (devillove084) [reply src/main.rs]: done

File: src/main.rs [hunk 21]
```diff
@@ -103,205 +109,216 @@ pub enum RunAsRole {
...
 }
 
-/// Get arguments value
-#[allow(clippy::too_many_lines)] //allow for this function as there is no other logic in this function
-fn get_args(matches: &ArgMatches) -> CliArgs {
+/// Generate the default arg
+#[allow(clippy::too_many_lines)]
+//allow for this function as there is no other logic in this function
+#[must_use]
+fn get_default_arg_map() -> HashMap<String, Arg<'static, 'static>> {
```
Comment (pzheng1025) [inline src/main.rs:194]: Arg<'static, 'static> --> Arg<'a, 'a>

File: src/main.rs [hunk 22]
```diff
@@ -402,89 +467,142 @@ fn get_args(matches: &ArgMatches) -> CliArgs {
...
+fn get_map<'a>(map: &HashMap<String, Arg<'a, 'a>>, name: &str) -> Arg<'a, 'a> {
+    if let Some(res) = map.get(name) {
+        res.clone()
+    } else {
+        panic!("Failed to get default argument {}", name);
     }
 }
 
+/// Parse the args in subcommand
+fn parse_args() -> ArgMatches<'static> {
```
Comment (pzheng1025) [inline src/main.rs:516]: ArgMatches<'static> --> ArgMatches<'a>

Comment (devillove084) [reply src/main.rs]: done

**Review by pzheng1025 — APPROVED** [review]

**Review by pwang7 — APPROVED** [review]

**Pull request merged**